### ISSUE TITLE
MODULE.bazel.lock file contains user specific paths

### ISSUE THREAD

**Issue by jsharpe** [bazelbuild/bazel#19621]
### Description of the bug:

If you use a workspace local registry then the attributes recorded in the lockfile contain the expanded value for %workspace% instead of using the %workspace% placeholder.

### Which category does this issue belong to?

External Dependency

### What's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Use `--registry=file://%workspace%/registry` and then depend on a module in that registry.

The applied patches from the registry are referenced using the expanded path of %workspace% instead of using a placeholder.

### Which operating system are you running Bazel on?

Ubuntu linux

### What is the output of `bazel info release`?

6.4.0rc1

**Comment by jsharpe** [comment]
@bazel-io flag

**Comment by iancha1992** [comment]
@bazel-io fork 6.4.0

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 6.4.0 RC2](https://releases.bazel.build/6.4.0/rc2/index.html). Please test out the release candidate and report any issues as soon as possible. Thanks!

**Comment by meteorcloudy** [comment]
@BoleynSu Can you please give more details of what's still broken? Ideally with a minimal reproducible case.

**Comment by Wyverald** [comment]
This is working as intended. If you're using a local path registry, we _have_ to write the path to the lockfile to ensure reproducibility (unless the path is under the workspace root). If you're just doing some local testing and don't want to commit things into the lockfile, please use `--lockfile_mode=off`.

**Comment by Wyverald** [comment]
@SalmaSamy could you try to repro and investigate? And close if unreproducible.

**Comment by SalmaSamy** [comment]
@BoleynSu Thanks, I was able to reproduce it and working on fixing it.

**Comment by fmeum** [comment]
Should this be cherry-picked into 7.0.1?

**Comment by Wyverald** [comment]
@bazel-io fork 7.1.0

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.1.0 RC1](https://releases.bazel.build/7.1.0/rc1/index.html). Please test out the release candidate and report any issues as soon as possible. Thanks!

**Comment by meteorcloudy** [comment]
@lalten What is your expected behaviour?

**Comment by lalten** [comment]
The `"path": "/home/username/monorepo/modules/rules_qnx"` part is not portable.
Probably something like `"path": "%workspace%/modules/rules_qnx"` would be better.
For now we cannot use the Lockfile because of the absolute path that's not present on most machines.

**Comment by meteorcloudy** [comment]
Oh, sorry, I missed that.
 
```
      "repoSpec": {
        "ruleClassName": "local_repository",
        "attributes": {
          "path": "/home/username/monorepo/modules/rules_qnx"
        }
      }
```
But why is the root module a `local_repository`? /cc @SalmaSamy @Wyverald 

@lalten Can you create a full reproducible case for this?

**Comment by appthumb** [comment]
I'm having the same issue, when I use my local registry relative to my %workspace%, and the MODULE.bazel.lock ends up storing the full absolute path in the "attribute".

The directory structure:

root
    |--- bazel_registry / modules / repo_A
    |--- repo_A
    |--- repo_B

And repo_B depends on repo_A through `bazel_registry`.

In `repo_B/.bazelrc`, I have:

```
common --registry=file://%workspace%/../bazel_registry
common --registry=https://bcr.bazel.build
```

In `bazel_registry/bazel_registry.json`, I have:

```
{
  "module_base_path": "../"
}
```

In `bazel_registry/modules/repo_A/1.0.0/source.json`, I have:

```
{
  "type": "local_path",
  "path": "repo_A"
}
```

This works correctly locally, and Bazel correctly finds the /root/repo_A and uses it when I build repo_B.

But the MODULE.bazel.lock file contains the absolute full path of repo_A in the "attribute":

```
      "repoSpec": {
        "ruleClassName": "local_repository",
        "attributes": {
          "path": "/Users/<my-local-directory>/repo_A"
        }
      }
```

This makes it not portable on a different machine, because the local path is different. Ideally the resolved attributes would be relative to %workspace%. According to [bazel document](https://bazel.build/external/registry#index_registry), the source path would be resolved as:

> If path and module_base_path are both relative paths, it resolves to <registry_path>/<module_base_path>/<path>. Registry must be hosted locally and used by --registry=file://<registry_path>

So something like "%workspace%/../bazel_registry/../repo_A" would be a direct translation of the rules. Simplifying it to "%workspace%/../repo_A" also works.

**Comment by appthumb** [comment]
Also, I'm using bazel 7.1.1.

```
$ bazelisk --version
bazel 7.1.1
```

**Comment by meteorcloudy** [comment]
@SalmaSamy Can you take a look?

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.2.0 RC1](https://github.com/bazelbuild/bazel/releases/tag/7.2.0rc1). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting USE_BAZEL_VERSION=7.2.0rc1. Thanks!

**Comment by appthumb** [comment]
Thanks for the fix! It works smoothly with my local bazel registry, and the MODULE.bazel.lock doesn't include any local path and works fine across different workspaces. Interestingly, when I look at the lock file, the dependency on my local repo is not mentioned at all, which seems to be intended according to https://github.com/bazelbuild/bazel/issues/19621#issuecomment-2079621790.

Overall this is very nice, thanks!
